### ISSUE TITLE
Could it be possible to be Wayland compatible?

### ISSUE THREAD

**Issue by q2dg** [phil294/AHK_X11#2]
Sorry if it's a silly question.

**Comment by phil294** [comment]
Not a silly question at all. Currently, all Hotkey logic etc. is heavily based on Xorg, so it would significant work, if possible at all. But I guess it will (have to) at some point.

I'm clueless about Wayland, I guess this thread provides some useful info though https://news.ycombinator.com/item?id=32654500

**Comment by aaaa2aaaa** [comment]
For wayland, it currently depends on whether or not the compositors have implemented protocols to do some of these actions that are considered a security risk. Currently I think only wlroots-based compositors let you do this. In the future however, this kind of functionality would be permitted through the use of [portals](https://github.com/flatpak/xdg-desktop-portal) which would ask the user to allow permission first.

So basically you would either have to use wlroots specific compositor protocols which might be removed in the future or give your support for the ability to do this in portals.

Currently I think there's portals for input/reading input in development but you would have to open issues for other functionality (and you may have to argue with people to convince them that the functionality is needed)

This might sound ridiculous since on xorg this stuff "just works" but at least when it's done only programs you choose can manipulate input and do all this other stuff when you give them permission to do it.

Hope this helps

**Comment by phil294** [comment]
if anybody wants to give it a go, [espanso](https://github.com/espanso/espanso) seems to have solved these problems (?). I think anything that works there, should be applicable to this project too.

**Comment by aaaa2aaaa** [comment]
That requires root access and you can only read input and write input, a very small subset of AHK's functionality.

**Comment by aaaa2aaaa** [comment]
Currently the only reasonable way to implement something like AHK on wayland would be to make your own wayland protocols then add support for them in compositors then compile the compositors yourself

**Comment by phil294** [comment]
oh so it does require root, okay. Thanks for your input, this will surely come handy some day.

**Comment by jf2048** [comment]
>In the future however, this kind of functionality would be permitted through the use of [portals](https://github.com/flatpak/xdg-desktop-portal) which would ask the user to allow permission first.

The project to get input emulation working with portals is called libei, you can see some blog posts about it here:

https://who-t.blogspot.com/2020/08/libei-library-to-support-emulated-input.html

https://who-t.blogspot.com/2021/08/libei-status-update.html

http://who-t.blogspot.com/2022/03/libei-adding-support-for-passive.html

Those should have links to some of the portal PRs.

>Currently the only reasonable way to properly implement something like AHK on wayland would be to make your own wayland protocols then add support for them in compositors then compile the compositors yourself

Well no, an interested party should give feedback on the drafts for the portal and help get those finished and stabilized. Hacking the compositors is unlikely to fix anything, as the real issue is how there is a lack of good API.

**Comment by aaaa2aaaa** [comment]
AHK doesn't just do input emulation, it also has stuff like manipulating windows. Is there any portal drafted to do that stuff? Right now I don't think there's even a standardized way to see the coordinates of a window.

**Comment by jf2048** [comment]
I have not seen any proposals for that. Someone may attempt to do it at some point, but my personal opinion is that efforts to do so will fail. Window managers on Linux are just too fragmented, even something as innocuous-seeming as window coordinates is unreliable and means different things depending on the window manager. This is a problem even on X11...

**Comment by phil294** [comment]
I have read a few bits about this now and wow, the situation here is dire. I think the sanest way is to go the evdev / uinput route here.

Since executables compiled by ahk_x11 *must* be independent and portable, it cannot depend on any kind of managed background service, installer script or manual setup. Instead, I suggest that every script *that uses any related feature* and runs on a non-x11 system spawns its own little child process and exchanges data via socket. That process attempts to hook into the raw keyboard events at script start. If it fails to do so, the end user is prompted to choose from three buttons:

> Insufficient permission to read your keyboard (Wayland). Please choose:
> 
> 1. [**Recommended**] Insert password to continue: Needs to be done for every program start
> 2. Re-run the script with root permissions: Not recommended
> 3. Add user to `input` group: Insecure and potentially dangerous, not recommended, but solves this for all future scripts

Those are essentially one-click actions and run nicely interactive with `sudo --askpass`. It gives non-technical users a recommendation while being clear and configurable to power users.

From what I have seen espanso and ydotool do, uinput code is no rocket science (in contrast to X11-targetting code), so this would be a clear path towards a realistic goal of supporting any graphical system.

All this display server specific logic like hotkeys, window management, key automation needs to be extracted into dedicated DisplayAdapter classes. This would also open up the possibility to add Windows support, but there is really no point in doing so.

**Comment by phil294** [comment]
FYI I'm working on this currently but it is an awful lot of work unfortunately, so I don't know yet when it will be finished. A lot of stuff seems to work, in parts even better than on X11, and others not at all, esp. Win* commands.

**Comment by aaaa2aaaa** [comment]
It'll probably be a while before you can even get some win* commands working. Look how long this protocol has been stuck:
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/59
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/155
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/187

**Comment by phil294** [comment]
oh wow, but at least something is happening. It's admittedly a bit hard to follow though.

Yes, I'm not even taking that route. But thanks to AtSpi, the accessibility protocol, it seems that matching windows by name and pid is still possible, as is interacting with elements inside windows, so it looks like e.g. `ControlSend, , keys, Firefox` still works. But window matching options `A` and `ahk_class` are gone.

**Comment by phil294** [comment]
Hi!

[here is a first demo version of ahk_x11 with Wayland support](https://github.com/phil294/AHK_X11/releases/tag/0.5.99-next). I don't really recommend yet to use it, but if you feel adventurous, you can now. Several things don't work properly yet, so check out the Readme from the Wayland branch https://github.com/phil294/AHK_X11/tree/evdev-wayland#wayland-evdev

I've figured that at-spi, the a11y protocol under Linux, is more capable than I thought and I need to read up onto a lot of things. This will still take a while. I don't have super much time available currently, so I can't tell for sure when I can continue to work on all of this. Because of that, I'm simply sharing what I have built so far.

**Comment by mercifulboss** [comment]
In KDE it seems to be possible to get window coordinates, and various other properties on Wayland when establishing window rules. In fact quite a lot of information can be obtained abt windows, including titles, children, etc. perhaps this might be something to look into?

**Comment by phil294** [comment]
@mercifulboss is this compositor-specific api? because I'd like to stay in wayland land as much as possible and avoid implementing things for all compositors in parallel.

**Comment by mercifulboss** [comment]
> @mercifulboss is this compositor-specific api? because I'd like to stay in wayland land as much as possible and avoid implementing things for all compositors in parallel.

This is compositor independent api.

My understanding (and I may be wrong) is that it is a daemon that creates a virtual keyboard using DBUS which will both detect keypresses and overwrites text when it detects certain user defined (in a configuration file) keywords. In this way it entirely bypasses the compositor. Again, this is my very basic understanding of how this works

**Comment by phil294** [comment]
fyi I've given up on this (and on Wayland in general. [It sucks.](https://gist.github.com/probonopd/9feb7c20257af5dd915e3a9f2d1f2277)). And I have never received any notable interest in it for ahk_x11 outside of this very issue anyway. If anybody should ever feel like tackling this, the POC branch is linked above, including lots of comments in the source and information above by commenters. I'll be happy to merge something that adds proper support, given that it doesn't break any existing x11 functionality.

tl;dr: Just use X11.

**Comment by jf2048** [comment]
That page is a lot of misinformation and conspiracy theories, you can freely ignore everything written there. Or see the various responses, for example: https://pointieststick.com/2023/12/26/does-wayland-really-break-everything/

Anyway I hope someone can continue that branch, since it uses AT-SPI it should work as a backend on both X11 and Wayland, and has the potential to be an all-around improvement.

**Comment by Descolada** [comment]
@jf2048 here is a non-exhaustive list of some features that AHK has:

1. Window management: move, minimize, restore, maximize, access the current state and position
2. Window controls: enumerate UI elements (buttons, textboxes etc), access element information (location, text, other metadata like class, id etc), control the elements (eg programmatically trigger button clicks, select listbox items, set text fields)
3. Screen capture: AHK is able to capture the whole screen for `ImageSearch`, but a limited version to capture only the active window or a specific window would be almost as useful. Capturing the raw image is not built-in AHK, but there are libraries written to do that via GDIP. 
4. Keyboard input-output: this doesn't just mean registering global hotkeys, but the ability to detect all keyboard input and selectively block input or let it through, as well as generate artificial input. In Windows detecting keyboard input is limited to apps with the same elevation level (low-level hooks don't call their callbacks if the active window is elevated but AHK isn't), and low-level keyboard hook doesn't provide device information (hopefully this can be different in Linux). Additionally, sending input is mostly limited to the active window. There is `ControlSend` which posts keystroke messages directly to the window, but the success rate of it working is rather low. 
5. Mouse input-output: same as above. Needs to detect absolute mouse position, detect mouse movements, block them, and generate artificial mouse movements and clicks.

All this is done without external packages/libraries, and with user-level permissions. 

How much of this list is doable with AT-SPI?
